### DISCUSSION TITLE
Can't get composables to work

### DISCUSSION THREAD

**Discussion by phillipstemann** [nuxt/nuxt#16112] [Questions]
Hi

I'm playing around with composables, but I can't get it to work.

composable/user.ts
```
import { useState } from '#app'

export default function () {
  return 'bar'
}
```

pages/index.vue
```
<script>
export default {
  layout: "login",
  async setup() {
    const user = user()
    return user
  },
  mounted() {
  },
  beforeMount() {
    feather.replace()
  }
};
</script>
```

I get the following error:
index.vue:169 Uncaught (in promise) ReferenceError: Cannot access 'user' before initialization

**Comment by wobsoriano** [comment]
For default exports, you should name your file like one of the ff:

1. `use-user.ts`
2. `useUser.ts`

```html
<script>
export default {
  setup() {
    const user = useUser()
    return {
      user
    }
  }
};
</script>
```

More info https://v3.nuxtjs.org/docs/directory-structure/composables/

**Reply by phillipstemann** [reply]
Thank you for taking your time to help me wobsoriano, I really appreciate it.
However, I get a new error now: Unhandled Promise Rejection: ReferenceError: Can't find variable: useUser

I have renamed the file to use-user.ts
And then changed my setup function to exactly what you added in your code.

The setup function is executed in pages/index.vue - Does that make a difference?
Is there something I'm, missing to install?

**Reply by phillipstemann** [reply]
![Screenshot 2021-11-02 at 15 14 09](https://user-images.githubusercontent.com/50196418/139864350-d02c617a-276d-4994-9c02-5daca3ad6cbd.jpg)
I've taken a picture of the setup so you can see it exactly, I hope you can spot what I'm doing wrong. Again, thank you so much for helping.

**Reply by phillipstemann** [reply]
I don't know if it has anything to do with it, but when I check .nuxt/auto-import.d.ts then useUser is not present in there. Can that be the issue, and if so how do I solve it?

**Reply by wobsoriano** [reply]
Try to delete `node_modules` and `.nuxt`, run `npm install` again

**Reply by phillipstemann** [reply]
Tried, it didn't work. Is it because I need to install @vue/composition-api?

**Reply by wobsoriano** [reply]
You don't need that

**Reply by wobsoriano** [reply]
wait.. are you in nuxt3 or bridge?

**Reply by phillipstemann** [reply]
I'm on nuxt3.
I've been using yarn, now I tried to change to npm after deleting the folders you mention, and then I get this error before it can render:
 WARN  [worker] Cannot find module '/Users/stemann/Developer/kolekti/front-end/node_modules/vue/server-renderer' imported from /Users/stemann/Developer/kolekti/front-end/.nuxt/nitro/index.mjs                                                                                                                                                                             14:01:36

**Reply by wobsoriano** [reply]
Hmm try to create a minimal reproduction using codesandbox or stackblitz

**Reply by phillipstemann** [reply]
It works on Codesandbox. Hmm, can it be my Node version? v17.0.1

https://codesandbox.io/s/cocky-hooks-ig6ns?file=/pages/index.vue

**Reply by phillipstemann** [reply]
It works on my local now as well, this is so strange. But thank you so much for your help wobsoriano. And sorry for the confusion, I don't understand why it works now.

**Reply by wobsoriano** [reply]
No worries. Glad it's working now!

**Reply by phillipstemann** [reply]
@wobsoriano How do I use these functions inside methods?

```
<script>
export default {
  setup() {
    const { user, signIn } = useUser()
    return {
      user,
      signIn
   }
  },
  layout: "login",
  mounted() {
  },
  beforeMount() {
    feather.replace()
  },
  methods: {
    signUserIn() {
      console.log("Jeps")
      this.signIn('Phillip')
    }
  }
};
</script>
```

I get the following error: Unhandled Promise Rejection: Error: nuxt instance unavailable

Here is the signIn function:
 ```
 const signIn = async (user) => {
    const { data, error } = await useFetch('http://localhost:5000/login', {
      method: 'post',
      headers: {
        Authorization: 'Bearer'
      },
      body: {
        email: 'test@test.dk',
        password: 'test'
      }
    })
    console.log(data)
    console.log(error._value)
    user.value = user
  }
```

**Reply by danielroe** [reply]
`useFetch` does something rather different from what you're expecting (it has to be called in the `setup()` function). You should likely just use `$fetch`:
```js
const signIn = async (user) => {
   user.value = await $fetch('http://localhost:5000/login', {
     method: 'post',
     headers: {
       Authorization: 'Bearer'
     },
     body: {
       email: 'test@test.dk',
       password: 'test'
     }
   })
 }
```

**Reply by phillipstemann** [reply]
Thank you so much Daniel, it's working.

**Comment by radekz2** [comment]
Posting this in case somebody lands here and is still stuck after Nuxt 3 stable release. I have ran into the same issue, but the root cause had to do with variable naming in component, see below.

My composable which groups globlal state vars, solution found here https://nuxt.com/docs/getting-started/state-management#shared-state

```TypeScript
export const useShowPanelEditCampaignName = () => useState<boolean>('useShowPanelEditCampaignName', () => false)
export const useShowPanelAddChannel = () => useState<boolean>('useShowPanelAddChannel', () => false)
```

In my component I would try to use those state values as follows:

```TypeScript
<script setup>
import { Menu, MenuButton, MenuItem, MenuItems } from '@headlessui/vue'
import { ChevronDownIcon } from '@heroicons/vue/20/solid'

const props = defineProps({
  campaign: Object
})

const showPanelEditCampaignName = useShowPanelEditCampaignName()
const useShowPanelAddChannel = useShowPanelAddChannel()
const route = useRoute()

</script>
```

And would get the following exception `Cannot access 'useShowPanelAddChannel' before initialization`

The problem was with the name of the second variable, it cannot be `useShowPanelAddChannel` as it matched the name of the composable, renaming to `showPanelAddChannel` resolved the issue.
